### PR TITLE
GODRIVER-1460 document mgocompat

### PR DESCRIPTION
**Pull request by iwysiu**

### CODE REVIEW

File: bson/mgocompat/doc.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+//
+// 8) The driver's primitive.Binary is equivalent to bson.Binary, with variables named Subtype
+//    and Data instead of Kind and Data.
+//
+// 9) The driver's primitive.Regex is equivalent to mgo's bson.RegEx.
+//
+// 10) The driver's primitive.JavaScript is equivalent to mgo's bson.JavaScript with no
+//     scope and primitive.CodeWithScope is equivalent to mgo's bson.JavaScript with scope.
+//
+// 11) The driver's primitive.DBPointer is equivalent to bson.DBPointer, with variables
```
Comment (divjotarora) [inline bson/mgocompat/doc.go]: Change to `is equivalent to mgo's bson.DBPointer` for consistency.

File: bson/mgocompat/doc.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+//     named DB and Pointer instead of Namespace and Id.
+//
+// 12) When implementing the setter interface, mgocompat.ErrSetZero is equivalent to mgo's
```
Comment (divjotarora) [inline bson/mgocompat/doc.go]: Capitalize `setter`.

File: bson/mgocompat/doc.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+//     bson.ErrSetZero.
+//
+// Note that if a value matches the hook for bsoncodec.Marshaler, bsoncodec.ValueMarshaler, or
```
Comment (divjotarora) [inline bson/mgocompat/doc.go]: This seems out of place. I think we should move it to the top of the file where we introduce the new registries and say that they register hooks for `Getter`/`Setter`. After that, say that those hooks are registered after the ones for `Marshaler`/`ValueMarshaler`/`Proxy` so those take priority if a type implements more than one of those interfaces.

File: bson/mgocompat/doc.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+// The driver uses different types than mgo's bson. The differences are:
+//
+// 1) The driver's bson.RawValue is equivalent to mgo's bson.Raw, but uses Value instead
+//    of Data and uses Type, which is a bsontype.Type object that wraps a byte, instead of
+//    bson.Raw's Kind, a byte.
+//
+// 3) The driver uses primitive.ObjectID, which is a [12]byte instead of mgo's
+//    bson.ObjectId, a string. Due to this, the zero value marshals and unmarshals differently
+//    for Extended JSON, with the driver marshaling as `{"ID":"000000000000000000000000"}` and
+//    mgo as `{"Id":""}`. The driver will not unmarshal {"ID":""} or `{"Id":null}` to a
```
Comment (divjotarora) [inline bson/mgocompat/doc.go]: `{ID: null}` should work after GODRIVER-1175.

File: bson/mgocompat/doc.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+// remaining differences. For compatibility with mgo's bson with RespectNilValues set to true,
+// use MgoRegistryRespectNilValues. See the bsoncodec docs for more details on registries.
+//
+// The differences between mgocompat and mgo's bson are:
+//
+// 1) mgocompat errors instead of silently skipping mismatched types when decoding.
+//
+// 2) mgocompat does not have special handling for marshaling array ops ("$in", "$nin", "$all").
+//
+// 4) mgocompat will error if you try to unmarshal an empty raw.
```
[the remark this answers is not in the document]
Comment (divjotarora) [reply bson/mgocompat/doc.go]: I don't recall making any changes for `Raw` in that PR. Can you show the test here?

File: bson/mgocompat/doc.go (hunk 4 above)
Comment (iwysiu) [reply bson/mgocompat/doc.go]: done

File: bson/mgocompat/doc.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (iwysiu) [reply bson/mgocompat/doc.go]: Talked offline and decided not to do/say anything about different bson.RawValue vs mgo bson.Raw behavior

File: bson/mgocompat/doc.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+// with some remaining differences. It also provides MgoRegistryRespectNilValues for compatibility
+// with mgo's BSON with RespectNilValues set to true. See the bsoncodec docs for more details on
+// registries.
+//
+// MgoRegistry supports Getter and Setter equivalents by registering hooks. Note that if a value
+// matches the hook for bsoncodec.Marshaler, bsoncodec.ValueMarshaler, or bsoncodec.Proxy, that
+// hook will take priority over the Getter hook. The same is true for the hooks for
+// bsoncodec.Unmarshaler and bsoncodec.ValueUnmarshaler and the Setter hook.
+//
+// The differences between MgoRegistry and globalsign/mgo's BSON library are:
```
Comment (jyemin) [inline bson/mgocompat/doc.go]: Consider "the functional differences"  here, since type differences are also a kind of difference.

Comment (iwysiu) [reply bson/mgocompat/doc.go]: done

File: bson/mgocompat/doc.go [hunk 7]
```diff
@@ -1,14 +1,14 @@
 // Package mgocompat provides MgoRegistry, a BSON registry compatible with globalsign/mgo's BSON,
 // with some remaining differences. It also provides MgoRegistryRespectNilValues for compatibility
-// with mgo's BSON with RespectNilValues set to true. See the bsoncodec docs for more details on
-// registries.
+// with mgo's BSON with RespectNilValues set to true. A registry can be configured on a
+// mongo.Client with the SetRegistry option. See the bsoncodec docs for more details on registries.
```
Comment (jyemin) [inline bson/mgocompat/doc.go]: Any further updates to this can be made in scope of https://jira.mongodb.org/browse/GODRIVER-1361

**Pull request merged**